src/Native/Unix/System.Security.Cryptography.Native/pal_x509.c

```diff
@@ -682,6 +688,126 @@ int32_t CryptoNative_X509StoreCtxCommitToChain(X509_STORE_CTX* storeCtx)
     return 1;
 }
 
+int32_t CryptoNative_X509StoreCtxResetForSignatureError(X509_STORE_CTX* storeCtx, X509_STORE** newStore)
+{
+    if (storeCtx == NULL || newStore == NULL)
+    {
+        return -1;
+    }
+
+    *newStore = NULL;
+
+    int errorDepth = X509_STORE_CTX_get_error_depth(storeCtx);
+    X509Stack* chain = X509_STORE_CTX_get0_chain(storeCtx);
+    int chainLength = sk_X509_num(chain);
+    X509_STORE* store = X509_STORE_CTX_get0_store(storeCtx);
+
+    // If the signature error was reported at the last element
+    if (chainLength - 1 == errorDepth)
+    {
+        X509* root;
+        X509* last = sk_X509_value(chain, errorDepth);
+
+        // If the last element is in the trust store we need to build a new trust store.
+        if (X509_STORE_CTX_get1_issuer(&root, storeCtx, last))
+        {
+            if (root == last)
+            {
+                // We know it's a non-zero refcount after this because last has one, too.
+                // So go ahead and undo the get1.
+                X509_free(root);
```

Comment:
This looks strange.  We're freeing something that's part of / still referenced by the store?

Reply:
> Or the get1_issuer call incremented the ref count and we're just undoing that?

Yeah, that.  `get0` means "I'm giving it to you without up_ref", `get1` means "I'm giving it to you, and it was up_ref'd"